source/Calamari.Shared/Integration/Packages/Download/S3PackageDownloader.cs

```diff
@@ -0,0 +1,131 @@
+﻿using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Amazon.Runtime;
+using Amazon.S3;
+using Amazon.S3.Model;
+using Calamari.Common.Commands;
+using Calamari.Common.Features.Packages;
+using Calamari.Common.Plumbing.FileSystem;
+using Calamari.Common.Plumbing.Logging;
+using Octopus.CoreUtilities.Extensions;
+using Octopus.Versioning;
+
+namespace Calamari.Integration.Packages.Download
+{
+    public class S3PackageDownloader : IPackageDownloader
+    {
+        static string[] knownFileExtensions = { ".zip", ".tar.gz", ".tar.bz2", ".tar.gz", ".tgz", ".tar.bz" };
+        
+        const char BucketFileSeparator = '/';
+        readonly ILog log;
+        readonly ICalamariFileSystem fileSystem;
+        
+        static readonly IPackageDownloaderUtils PackageDownloaderUtils = new PackageDownloaderUtils();
+
+        public S3PackageDownloader(ILog log, ICalamariFileSystem fileSystem)
+        {
+            this.log = log;
+            this.fileSystem = fileSystem;
+        }
+        
+        string BuildFileName(string prefix, string version, string extension)
+        {
+            return $"{prefix}.{version}{extension}";
+        }
+        
+        (string BucketName, string Filename) GetBucketAndKey(string searchTerm)
+        {
+            var splitString = searchTerm.Split(new [] { BucketFileSeparator }, 2);
+            if (splitString.Length == 0)
+                return ("", "");
+            if (splitString.Length == 1)
+                return (splitString[0], "");
+            
+            return (splitString[0], splitString[1]);
+        }
+
+        public PackagePhysicalFileMetadata DownloadPackage(string packageId,
+                                                           IVersion version,
+                                                           string feedId,
+                                                           Uri feedUri,
+                                                           string? feedUsername,
+                                                           string? feedPassword,
+                                                           bool forcePackageDownload,
+                                                           int maxDownloadAttempts,
+                                                           TimeSpan downloadAttemptBackoff)
+        {
+            var (bucketName, prefix) = GetBucketAndKey(packageId);
+            if (string.IsNullOrWhiteSpace(bucketName) || string.IsNullOrWhiteSpace(prefix))
+            {
+                throw new InvalidOperationException("Invalid PackageId for S3 feed. Expecting format `<bucketName>/<packageId>`");
```

Comment:
Should we put what was received in the exception too? i.e. `received {packageId}, expected format ...`